### PR TITLE
feat(xds): Add ScaleWay service discovery to the ScrapeConfig CRD

### PR DESCRIPTION
**Pull request by dongjiang1989**
## Description

_Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request.
If it fixes a bug or resolves a feature request, be sure to link to that issue._

Add ScaleWay service discovery to the ScrapeConfig CRD.

ref: https://github.com/prometheus-operator/prometheus-operator/pull/6634

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [x] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Verification

Please check the [Prometheus-Operator testing guidelines](../TESTING.md) for recommendations about automated tests.

## Changelog entry

_Please put a one-line changelog entry below. This will be copied to the changelog file during the release process._

```release-note
Add ScaleWay service discovery to the ScrapeConfig CRD
```

### CODE REVIEW

**Comment by mviswanathsai** [comment]
Thanks for the PR! I will check it out today or by Monday at the earliest.

File: pkg/prometheus/promcfg.go [hunk 1]
```diff
@@ -4044,6 +4043,100 @@ func (cg *ConfigGenerator) generateScrapeConfig(
...
+			if config.RefreshInterval != nil {
+				configs[i] = append(configs[i], yaml.MapItem{
+					Key:   "refresh_interval",
+					Value: config.RefreshInterval,
+				})
+			}
+
+			configs[i] = cg.addProxyConfigtoYaml(configs[i], s, config.ProxyConfig)
+
+			if config.TLSConfig != nil {
```
Comment (mviswanathsai) [inline pkg/prometheus/promcfg.go:4115]: Could we do the same thing that we do with `ProxyConfig` etc,. and just handle the nil value of `TLSConfig` within the `addSafeTLSToYaml`? @simonpasquier @slashpai

File: pkg/prometheus/resource_selector.go [hunk 2]
```diff
@@ -1444,3 +1449,20 @@ func (rs *ResourceSelector) validateOVHCloudSDConfigs(ctx context.Context, sc *m
...
 	return nil
 }
+
+func (rs *ResourceSelector) validateScaleWaySDConfigs(ctx context.Context, sc *monitoringv1alpha1.ScrapeConfig) error {
+	promVersion := operator.StringValOrDefault(rs.p.GetCommonPrometheusFields().Version, operator.DefaultPrometheusVersion)
+	version, err := semver.ParseTolerant(promVersion)
+	if err != nil {
+		return fmt.Errorf("failed to parse Prometheus version: %w", err)
+	}
+	if !version.GTE(semver.MustParse("2.27.0")) {
```
Comment (mviswanathsai) [inline pkg/prometheus/resource_selector.go:1459]:
```suggestion
	if !version.GTE(semver.MustParse("2.26.0")) {
```

File: pkg/prometheus/promcfg_test.go [hunk 3]
```diff
@@ -9963,6 +9963,189 @@ func TestScrapeConfigSpecConfigWithOVHCloudSD(t *testing.T) {
...
+								},
+								Key: "private-key",
+							},
+						},
+					},
+				},
+			},
+			golden: "ScrapeConfigSpecConfig_ScaleWaySD.golden",
+		}, {
+			name: "scaleway_sd_config_tls",
```
Comment (mviswanathsai) [inline pkg/prometheus/promcfg_test.go:10038]: Since TLS is already covered in the first test, can we remove this if we are not testing for anything new?

**Review by mviswanathsai — COMMENTED** [review]
Just a few small comments, looks good overall.

File: pkg/prometheus/promcfg.go [hunk 4]
```diff
@@ -4044,6 +4043,100 @@ func (cg *ConfigGenerator) generateScrapeConfig(
...
+			if config.EnableHTTP2 != nil {
+				configs[i] = append(configs[i], yaml.MapItem{
+					Key:   "enable_http2",
+					Value: config.EnableHTTP2,
+				})
+			}
+		}
+
+		cfg = append(cfg, yaml.MapItem{
+			Key:   "scaleway_sd_config",
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:4135]:
```suggestion
			Key:   "scaleway_sd_configs",
```

Comment (dongjiang1989) [reply pkg/prometheus/promcfg.go]: Thanks. Done.

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go [hunk 5]
```diff
@@ -1168,3 +1171,65 @@ type LightSailSDConfig struct {
...
+	// +kubebuilder:validation:MinLength=1
+	// +required
+	AccessKey string `json:"accessKey"`
+	// Secret key to use when listing targets.
+	// +required
+	SecretKey corev1.SecretKeySelector `json:"secretKey"`
+	// Project ID of the targets.
+	// +required
+	ProjectID string `json:"projectID"`
+	// Service of the targets to retrieve. Must be `VPS` or `DedicatedServer`.
```
[the remark this answers is not in the document]
Comment (dongjiang1989) [reply pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go]: My bad. Done. Thanks.

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go [hunk 6]
```diff
@@ -1168,3 +1171,65 @@ type LightSailSDConfig struct {
...
+
+const (
+	Instance  ScaleWayRole = "Instance"
+	Baremetal ScaleWayRole = "Baremetal"
+)
+
+// ScaleWaySDConfig configurations allow retrieving scrape targets from Scaleway instances and baremetal services.
+// See https://prometheus.io/docs/prometheus/latest/configuration/configuration/#scaleway_sd_config
+// TODO: Need to document that we will not be supporting the `_file` fields.
+type ScaleWaySDConfig struct {
```
[the remark this answers is not in the document]
Comment (dongjiang1989) [reply pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go]: Thanks. Done.

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go [hunk 7]
```diff
@@ -1168,3 +1171,65 @@ type LightSailSDConfig struct {
...
 }
+
+// Role of the targets to retrieve. Must be `Instance` or `Baremetal`.
+// +kubebuilder:validation:Enum=Instance;Baremetal
+type ScaleWayRole string
+
+const (
+	Instance  ScaleWayRole = "Instance"
+	Baremetal ScaleWayRole = "Baremetal"
+)
```
[the remark this answers is not in the document]
Comment (dongjiang1989) [reply pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go]: Thanks. Done.

File: pkg/prometheus/promcfg_test.go (hunk 3 above)
Comment (dongjiang1989) [reply pkg/prometheus/promcfg_test.go:10014]: Thanks. Remove it Done.

File: pkg/prometheus/resource_selector.go (hunk 2 above)
Comment (dongjiang1989) [reply pkg/prometheus/resource_selector.go]: My Bad. Fixed it Done.

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go [hunk 8]
```diff
@@ -1168,3 +1171,65 @@ type LightSailSDConfig struct {
...
+	// +optional
+	NameFilter *string `json:"nameFilter,omitempty"`
+	// TagsFilter specify a tag filter (a server needs to have all defined tags to be listed) to apply on the server listing request.
+	// +optional
+	TagsFilter []string `json:"tagsFilter,omitempty"`
+	// Refresh interval to re-read the list of instances.
+	// +optional
+	RefreshInterval *v1.Duration `json:"refreshInterval,omitempty"`
+	// ProxyConfig allows customizing the proxy behaviour for this scrape config.
+	// +optional
```
[the remark this answers is not in the document]
Comment (dongjiang1989) [reply pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:1227]: remove it done.

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go [hunk 9]
```diff
@@ -1168,3 +1171,65 @@ type LightSailSDConfig struct {
...
+	ApiURL *string `json:"apiURL,omitempty"`
+	// Zone is the availability zone of your targets (e.g. fr-par-1).
+	// +optional
+	Zone *string `json:"zone,omitempty"`
+	// NameFilter specify a name filter (works as a LIKE) to apply on the server listing request.
+	// +optional
+	NameFilter *string `json:"nameFilter,omitempty"`
+	// TagsFilter specify a tag filter (a server needs to have all defined tags to be listed) to apply on the server listing request.
+	// +optional
```
[the remark this answers is not in the document]
Comment (dongjiang1989) [reply pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:1222]: Thanks. Add it done.

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go [hunk 10]
```diff
@@ -1168,3 +1171,65 @@ type LightSailSDConfig struct {
...
+	// +kubebuilder:validation:Minimum=0
+	// +kubebuilder:validation:Maximum=65535
+	// +optional
+	Port *int32 `json:"port,omitempty"`
+	// API URL to use when doing the server listing requests.
+	// +kubebuilder:validation:Pattern:="^http(s)?://.+$"
+	// +optional
+	ApiURL *string `json:"apiURL,omitempty"`
+	// Zone is the availability zone of your targets (e.g. fr-par-1).
+	// +optional
```
[the remark this answers is not in the document]
Comment (dongjiang1989) [reply pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:1214]: Thanks. Add it done.

File: pkg/prometheus/promcfg.go (hunk 1 above)
Comment (dongjiang1989) [reply pkg/prometheus/promcfg.go:4115]: I can optimize it in the next PR.

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go [hunk 11]
```diff
@@ -1168,3 +1171,68 @@ type LightSailSDConfig struct {
 	// +optional
 	EnableHTTP2 *bool `json:"enableHTTP2,omitempty"`
 }
+
+// Role of the targets to retrieve. Must be `Instance` or `Baremetal`.
+// +kubebuilder:validation:Enum=Instance;Baremetal
```
[the remark this answers is not in the document]
Comment (dongjiang1989) [reply pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:1176]: Thanks. Remove it done.

**Review by simonpasquier — APPROVED** [review]
Thanks!

**Pull request merged**